core/src/main/java/com/linecorp/armeria/server/RouteBindingBuilders.java

```diff
@@ -0,0 +1,412 @@
+                for (HttpMethod method : newMethods) {
+                    if (!oldMethods.add(method)) {
+                        throw new IllegalArgumentException(
+                                "duplicate HTTP method: " + method + ", for: " + pathPatternBuilders);
+                    }
+                }
+                return oldMethods;
+            });
+        }
+
+        @Override
+        public AbstractRouteBindingBuilder methods(HttpMethod... methods) {
+            return methods(ImmutableSet.copyOf(requireNonNull(methods, "methods")));
+        }
+
+        @Override
+        public AbstractRouteBindingBuilder methods(Iterable<HttpMethod> methods) {
+            requireNonNull(methods, "methods");
+            checkArgument(!Iterables.isEmpty(methods), "methods can't be empty");
+            this.methods = ImmutableSet.copyOf(methods);
+            return this;
+        }
+
+        protected Set<HttpMethod> methods() {
+            return methods;
+        }
+
+        @Override
+        public AbstractRouteBindingBuilder consumes(MediaType... consumeTypes) {
+            consumes(ImmutableSet.copyOf(requireNonNull(consumeTypes, "consumeTypes")));
+            return this;
+        }
+
+        @Override
+        public AbstractRouteBindingBuilder consumes(Iterable<MediaType> consumeTypes) {
+            ensureUniqueMediaTypes(consumeTypes, "consumeTypes");
+            this.consumeTypes = ImmutableSet.copyOf(consumeTypes);
+            return this;
+        }
+
+        protected Set<MediaType> consumes() {
+            return consumeTypes;
+        }
+
+        @Override
+        public AbstractRouteBindingBuilder produces(MediaType... produceTypes) {
+            produces(ImmutableSet.copyOf(requireNonNull(produceTypes, "produceTypes")));
+            return this;
+        }
+
+        @Override
+        public AbstractRouteBindingBuilder produces(Iterable<MediaType> produceTypes) {
+            ensureUniqueMediaTypes(produceTypes, "produceTypes");
+            this.produceTypes = ImmutableSet.copyOf(produceTypes);
+            return this;
+        }
+
+        protected Set<MediaType> produces() {
+            return produceTypes;
+        }
+
+        protected List<Route> pathPatternRoutes() {
+            final ImmutableList.Builder<Route> builder = ImmutableList.builder();
+            pathPatternBuilders.forEach((pathPattern, methods) -> {
+                builder.add(Route.builder()
+                                 .path(pathPattern)
+                                 .methods(methods)
+                                 .consumes(consumes())
+                                 .produces(produces())
+                                 .build());
+            });
+            return builder.build();
+        }
+    }
+
+    /**
+     * A builder class for binding a {@link Route} with a {@code pathPattern} or a {@code prefix}.
+     */
+    static class OnePathRouteBindingBuilder extends AbstractRouteBindingBuilder {
```

Comment:
Maybe `SingeRouteBindingBuilder`?

Reply:
We chatted on this complicated logic and hierarchy.
To simplify this, I am going to remove `OnePathRouteBindingBuilder`, `MultiPathRouteBindingBuilder` and migrate only into `AbstractBindingBuilder`.